### ISSUE TITLE
Padding and evaluation details

### ISSUE THREAD

**Issue by VoiceBeer** [allenai/data-efficient-finetuning#1]
Hi, thank you for your brilliant work! It's fascinating and brings a lot of inspiration!

If I may, I have a few questions on training and evaluation details:
1. Did you use `padding` during training? I only found `padding=True` in _/scripts/index_p3_train_reps.py_ for indexing.
2. During training, the loss is calculated on the predicted token with the ground-truth token in _line 106_ from _/attribution/model.py_. No string matching, only ids matching, is that right? 
3. For evaluation, I was wondering that how you get the target token. I didn't see any code like `model.generate()` .
4. In both training and evaluation phases, only token ids are taken as the input and output, not string sentences right? 

Thx in advance, and looking for your future work xD

**Comment by hamishivi** [comment]
Hi! Sorry for the delay in answering, I just noticed this!

1. Yes, padding is applied during training. This is done by the default allennlp data collator, as it converts instances into Batch objects (see https://docs.allennlp.org/main/api/data/batch/)
2. This codepath actually isn't used during training - during training we instead just use the [loss from the underlying transfomers model](https://github.com/allenai/data-efficient-finetuning/blob/main/attribution/model.py#L81) (i.e., the loss calculated [here](https://github.com/huggingface/transformers/blob/main/src/transformers/models/t5/modeling_t5.py#L1751)). However, in both cases we calculate the loss using cross-entropy loss, comparing the model's output distribution with the ground truth token id.
3. For our evaluation, most tasks use rank classification evaluation, where we compute the loss of each possible answer and choose the model's prediction as the answer with lowest loss. This means we can measure model performance without worrying about output format. For Qasper, DROP, and supernatural instructions we do generate predictions using the allennlp evaluate command (which handles the model generation for us), e.g. [here](https://github.com/allenai/data-efficient-finetuning/blob/main/shell_scripts/evaluate_natural_instructions.sh). 
4. Correct, we have to tokenize and then convert the inputs and outputs to token ids before the model can process them. This is done in the dataset readers, e.g. [here](https://github.com/allenai/data-efficient-finetuning/blob/main/attribution/drop_reader.py#L127) for DROP.

I hope this is useful! Our codebase relies alot on the allennlp library for training and evaluation, so looking over the [allennlp documentation](https://docs.allennlp.org/main/) or [guide](https://guide.allennlp.org/) might be useful if you're unfamiliar with it.

**Comment by VoiceBeer** [comment]
No worries, It's just in time!

Thank you very much for your detailed answer, very helpful to me! I'll check out allennlp documentation in case I miss some details. Looking forward to your future work!

Good day 😆
